### ISSUE TITLE
Add all_of, any_of, and none_of

### ISSUE THREAD

**Issue by soonho-tri** [RobotLocomotion/drake#8504]
feature request by @jwnimmer-tri (from Slack channel)

```c++
bool is_any_of_nan(const Eigen::Ref<MatrixX<Expression>>&);
bool is_all_of_nan(const Eigen::Ref<MatrixX<Expression>>&);
```

**Comment by jwnimmer-tri** [comment]
I think we want them Formula-valued, and templated on T:
```c++
template <typename T>
drake::Bool<T> is_any_of_nan(const Eigen::Ref<MatrixX<T>>&);

template <typename T>
drake::Bool<T> is_all_of_nan(const Eigen::Ref<MatrixX<T>>&);
```

**Comment by soonho-tri** [comment]
I'm wondering where we need to put those functions. Do you have a suggestion?

**Comment by jwnimmer-tri** [comment]
I think `all_of` would likely only be used in unit tests -- to confirm that the "we guarantee to give you garbage" constructors actually do so.  Though in that case, `.array().isnan().all()` also works since we know there are no free variables.

**Comment by soonho-tri** [comment]
FYI, I'm adding more generic `any_of` and `all_of` which takes a matrix *and* a predicate.

**Comment by jwnimmer-tri** [comment]
New idea...

What if we mimic http://en.cppreference.com/w/cpp/algorithm/all_any_none_of with functions of that same name, but that take a `Matrix::Ref`, instead of iterators?  And place it in `drake_bool.h`?

**Comment by soonho-tri** [comment]
@jwnimmer-tri OK. I'll add them in `drake_bool.h`.

**Comment by soonho-tri** [comment]
Closed by #8508
